cot/src/db.rs

```diff
@@ -965,6 +1014,215 @@ impl Database {
         Ok(())
     }
 
+    /// Bulk inserts multiple rows into the database.
+    ///
+    /// # Errors
+    ///
+    /// This method can return an error if the rows could not be inserted into
+    /// the database, for instance because the migrations haven't been
+    /// applied, or there was a problem with the database connection.
+    pub async fn bulk_insert<T: Model>(&self, data: &mut [T]) -> Result<()> {
+        let span = span!(Level::TRACE, "bulk_insert", table = %T::TABLE_NAME, count = data.len());
+
+        Self::bulk_insert_impl(self, data, false)
+            .instrument(span)
+            .await
+    }
+
+    /// Bulk inserts multiple rows into the database, or updates them if they
+    /// already exist.
+    ///
+    /// # Errors
+    ///
+    /// This method can return an error if the rows could not be inserted into
+    /// the database, for instance because the migrations haven't been
+    /// applied, or there was a problem with the database connection.
+    pub async fn bulk_insert_or_update<T: Model>(&self, data: &mut [T]) -> Result<()> {
+        let span = span!(
+            Level::TRACE,
+            "bulk_insert_or_update",
+            table = %T::TABLE_NAME,
+            count = data.len()
+        );
+
+        Self::bulk_insert_impl(self, data, true)
+            .instrument(span)
+            .await
+    }
+
+    async fn bulk_insert_impl<T: Model>(&self, data: &mut [T], update: bool) -> Result<()> {
+        // TODO: add transactions when implemented
+
+        if data.is_empty() {
+            return Ok(());
+        }
+
+        let max_params = match self.inner {
+            // https://sqlite.org/limits.html#max_variable_number
+            // Assuming SQLite > 3.32.0 (2020-05-22)
+            #[cfg(feature = "sqlite")]
+            DatabaseImpl::Sqlite(_) => 32766,
+            // https://www.postgresql.org/docs/18/protocol-message-formats.html#PROTOCOL-MESSAGE-FORMATS-BIND
+            // The number of parameter format codes is Int16
+            #[cfg(feature = "postgres")]
+            DatabaseImpl::Postgres(_) => 65535,
+            // https://dev.mysql.com/doc/dev/mysql-server/9.5.0/page_protocol_com_stmt_prepare.html#sect_protocol_com_stmt_prepare_response
+            // The number of parameter returned in the COM_STMT_PREPARE_OK packet is int<2>
+            #[cfg(feature = "mysql")]
+            DatabaseImpl::MySql(_) => 65535,
+        };
```

Comment:
Not related to this PR, but this information should be returned by the database abstraction after the refactor.